sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -1532,6 +1532,13 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     assert(e.getMessage.contains("string, binary or array"))
   }
 
+  test("SPARK-31227: Non-nullable null type should not coerce to nullable type in concat") {
```

Comment:
How about `concat(array(), array(NULL))`? That should have the same type as `array(NULL)`.

Reply:
Yea, the output has the same type with array(null);
```
scala> sql("select concat(array(), array(NULL))").printSchema
root
 |-- concat(array(), array(NULL)): array (nullable = false)
 |    |-- element: null (containsNull = true)

scala> sql("select array()").printSchema
root
 |-- array(): array (nullable = false)
 |    |-- element: null (containsNull = false)

scala> sql("select array(null)").printSchema
root
 |-- array(NULL): array (nullable = false)
 |    |-- element: null (containsNull = true)
```
Any concern?

Reply:
Only that it should be tested, since it's an interesting corner case!

Reply:
I believe those cases are tested in `CastSuite.scala` and `TypeCoercionSuite.scala` including all types if I didn't miss anything. I just kept one e2e test here since it was the reported case in the JIRA SPARK-31227.